pkg/services/auth/internal/verifier.go

```diff
@@ -0,0 +1,44 @@
+package internal
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"github.com/weaveworks/weave-gitops/pkg/utils"
+	"strings"
+)
+
+type CodeVerifier struct {
+	min   int
+	max   int
+	value string
+}
+
+func NewCodeVerifier(min, max int) (CodeVerifier, error) {
+	value, err := utils.GenerateRandomString(min, max)
+	if err != nil {
+		return CodeVerifier{}, fmt.Errorf("new code verifier: %w", err)
+	}
+
+	return CodeVerifier{
+		min:   min,
+		max:   max,
+		value: value,
+	}, nil
+}
+
+func (c CodeVerifier) CodeChallenge() string {
+	h := sha256.New()
+	h.Write([]byte(c.value))
+
+	encodedValue := base64.StdEncoding.EncodeToString(h.Sum(nil))
+	encodedValue = strings.Replace(encodedValue, "+", "-", -1)
```

Comment:
Nope url will include them as well, such as replacing ' ' with '+'.  In addition, '=' is used as padding at the end of the encoding if the string isn't divisible by 3.  This is all part of the spec, the same removal is being done on the 3rd party server otherwise the code challenges won't match, nor will there conversion of the raw value.